src/query.rs

```diff
@@ -249,79 +251,88 @@ impl Query {
-
-                                    let cfg2 = Arc::clone(&cfg);
-                                    let query_state_holder2 = Arc::clone(&query_state_holder);
-
-                                    // Returns Result<(ds, files), error>. Need to stop on error.
-                                    // TODO: Stop on error
-                                    list_msl_bucket_files(log_name.as_str(), &ds)
-                                        .map(move |obj_key| {
-                                            (query_index.clone(), log_ds_index.clone(), obj_key)
-                                        })
-                                        .map_err(|e| QueryError::Underlying(format!("{:?}", e))) //temporarely remove error, we need to adress this
-                                        .map(move |(query_index, log_ds_index, obj_key)| {
-                                            // TODO: Limit the number of results
-                                            let read_state_holder =
-                                                query_state_holder2.read().unwrap();
-                                            let q_parse =
-                                                &read_state_holder.query_parsing[query_index].1;
-
-                                            let cfg_read = cfg2.read().unwrap();
-                                            let log = cfg_read.get_log(&q_parse.log_name).unwrap();
-
-                                            let ds_name = &log.datastores[log_ds_index];
-                                            let ds = cfg_read.datastore.get(ds_name).unwrap();
-
-                                            let query_state_holder2 =
-                                                Arc::clone(&query_state_holder2);
-
-                                            read_file_line_by_line(&obj_key, &ds)
-                                                .filter_map(move |line| {
-                                                    let read_state_holder =
-                                                        query_state_holder2.read().unwrap();
-                                                    let q = &read_state_holder.query_parsing
-                                                        [query_index]
-                                                        .0;
-                                                    let q_parse = &read_state_holder.query_parsing
-                                                        [query_index]
-                                                        .1;
-
-                                                    evaluate_query_on_line(&q, &q_parse, line)
-                                                })
-                                                .map_err(|e| {
+                            let query_state_holder3 = Arc::clone(&query_state_holder);
+
+                            let (tx, rx) = mpsc::unbounded_channel();
+
+                            for i in 0..logs_ds_len {
```

Comment:
Can we also do some code improvement in this PR? For example each task per data store can be in its own function.

Reply:
ready